# Pakistan Housing Market — Exploratory Data Analysis

An in-depth exploration of **16,000+ property listings** from [Zameen.com](https://www.zameen.com/), Pakistan's largest real estate marketplace. This analysis examines pricing patterns, geographic trends, and property characteristics across 12 major Pakistani cities to uncover actionable insights for real estate investors and market analysts.

**Dataset:** Zameen.com Housing Prices (Kaggle)  
**Tools:** Python, pandas, matplotlib, seaborn, plotly

## 1. Setup & Data Loading

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Visualization defaults
sns.set_style('whitegrid')
sns.set_palette('Set2')
plt.rcParams['figure.figsize'] = (10, 6)
plt.rcParams['figure.dpi'] = 150

pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:,.2f}'.format)

In [2]:
df = pd.read_csv('../data/raw/archive_2/zameen.csv', sep='|')
print(f'Dataset: {df.shape[0]:,} listings × {df.shape[1]} features')

Dataset: 16,044 listings × 6 features


## 2. Initial Exploration

### 2.1 Sample Listings

In [3]:
df.head(10)

city                                           location     price  \
0  Lahore                           DHA Phase 6, DHA Defence  74500000   
1  Lahore                           DHA Phase 7, DHA Defence  51500000   
2  Lahore                        Dream Gardens, Defence Road   7500000   
3  Lahore                           DHA Phase 6, DHA Defence  73000000   
4  Lahore                Bahria Town - Sector B, Bahria Town   5700000   
5  Lahore                 DHA Phase 5 - Block L, DHA Phase 5  53500000   
6  Lahore  Bahria Town - Overseas A, Bahria Town - Overse...  97500000   
7  Lahore  Bahria Town - Jasmine Block, Bahria Town - Sec...  47000000   
8  Lahore                Bahria Town - Sector E, Bahria Town   5000000   
9  Lahore                               Raiwind Road, Lahore   8299000   

   bedrooms  baths     size  
0         5      6 4,500.00  
1         5      6 4,500.00  
2         1      1   518.00  
3         5      6 4,500.00  
4         1      1   472.00  
5         5      6 2,250.00  
6         5      6 4,500.00  
7         5      7 4,500.00  
8         1      1   450.00  
9         1      1   630.00

### 2.2 Data Types & Memory

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 16044 entries, 0 to 16043
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   city      16044 non-null  str    
 1   location  16044 non-null  str    
 2   price     16044 non-null  int64  
 3   bedrooms  16044 non-null  int64  
 4   baths     16044 non-null  int64  
 5   size      16044 non-null  float64
dtypes: float64(1), int64(3), str(2)
memory usage: 752.2 KB


### 2.3 Numerical Summary

In [5]:
df.describe()

price  bedrooms     baths         size
count        16,044.00 16,044.00 16,044.00    16,044.00
mean     45,711,134.19      3.78      4.06     2,540.64
std      85,041,953.69      1.99      2.14    10,607.07
min          70,000.00      0.00      0.00         0.00
25%      14,000,000.00      3.00      3.00     1,125.00
50%      24,000,000.00      4.00      4.00     1,575.00
75%      45,000,000.00      5.00      6.00     2,700.00
max   2,100,000,000.00     11.00     10.00 1,215,000.00

### 2.4 Categorical Summary

In [6]:
df.describe(include='object')

C:\Users\Pc\AppData\Local\Temp\ipykernel_45676\87514550.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include='object')


city                 location
count    16044                    16044
unique      12                     2038
top     Lahore  DHA Villas, DHA Defence
freq      2500                      481

### 2.5 First Impressions

- **16,044 listings** across **12 cities** and **2,038 unique locations** — a broad snapshot of Pakistan's urban housing market
- **No missing values** in any column, though data quality issues exist (zero-size properties, extreme outliers)
- **Price range** spans 70,000 PKR to 2.1 Billion PKR — the high end likely includes commercial or incorrectly entered listings
- **Median price** (24M PKR) is roughly half the mean (45.7M PKR), indicating strong right-skew typical of real estate markets
- **Lahore dominates** with the highest listing count, followed by DHA/Bahria Town locations — Pakistan's premium housing societies
- **Size column** contains zero values and an extreme max of 1.2M sq ft — cleaning required before analysis